File: exer3/exer3.hs
import Data.Time.Calendar
import Data.Time.Calendar.OrdinalDate

merge :: Ord a => [a] -> [a] -> [a]
merge ([]) ([]) = []
merge x ([]) = x
merge ([]) y = y
merge (x:xs) (y:ys)
    | x >= y        = [y] ++ merge ([x] ++ xs) ys
    | x < y         = [x] ++ merge xs ([y] ++ ys)

mergeSort :: Ord a => [a] -> [a]
mergeSort ([]) = []
mergeSort (x:[]) = [x]
mergeSort list = merge (mergeSort a) (mergeSort b)
    where (a,b) = splitAt (length list `div` 2) list

daysInYear :: Integer -> [Day]
daysInYear y = [jan1..dec31]
  where jan1 = fromGregorian y 1 1
        dec31 = fromGregorian y 12 31

isFriday :: Day -> Bool
isFriday day = (weekDay == 5)
    where weekDay = snd (mondayStartWeek day)

divisors :: Int -> [Int]
divisors n = [i | i <- [2..(n `div` 2)], n `mod` i == 0]

isPrimeDay :: Day -> Bool
isPrimeDay day = (divisors dayNumber == [])
    where (year, month, dayNumber) = toGregorian day

primeFridays :: Integer -> [Day]
primeFridays y = [i | i <- daysInYear (y), isFriday (i), isPrimeDay(i)]

File: a2/exer3.erl
-module(exer3).
-export([merge/2, mergeSort/1, isPrimeDay/1, isFriday/1]).

merge([], []) -> [];
merge(X, []) -> X;
merge([], Y) -> Y;
merge([X|XS], [Y|YS]) ->
  case X >= Y of
    true    -> [Y] ++ merge([X] ++ XS, YS);
    false   -> [X] ++ merge(XS, [Y] ++ YS)
  end.

mergeSort([]) -> [];
mergeSort([X]) -> [X];
mergeSort(List) ->
  {A,B} = lists:split(length(List) div 2, List),
  merge(mergeSort(A), mergeSort(B)).

divisors(N) ->
  [I || I <- lists:seq(2, N div 2), N rem I =:= 0].

isPrimeDay(Date) ->
  {_, _, Day} = Date,
  divisors(Day) =:= [].

isFriday(Date) ->
  DayOfTheWeek = calendar:day_of_the_week(Date),
  DayOfTheWeek =:= 5.